File: src/main/java/com/craftmanship/DataValidator.java
package com.craftmanship;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Predicate;

// Vergleichbare andere Katas:
// Email Sender code refactroing kata
// error conditions refactroing kata?
// ValidateAndAddProduct-Refactoring-Kata

// Mögliche Änderungen:
// - neue Columns
// - Änderungen in der Reihenfolge der Columns
// - Änderungen/Fehler in Validierung
// - Mehrere Validierungen für ein Feld

// welches Design Möglichkeiten stehen zur Verfügung 
//  Liste von commands
//  Liste von Specifications
//  Monade
//  Composite Pattern
//  Template Method
//  Vavr Validation Framework?

//v1
//List<ErrorInfo> errors = new ValidatorCollector()
//	.validate(t -> Validators::validFirstName, "first name must contain characters", columns.get(0))
//	.validate(t -> Validators::validLastName, "last name must contain characters", columns.get(1))
//	.validate(t -> Validators::validXYZ, "last name must contain characters", columns.get(4))
//	.errors();
	
//v2	
//List<ValidationSpec> specs = List.of(
//	new ValidationSpec(t -> Validators::validFirstName, "first name must contain characters", columns.get(0))
//	new ValidationSpec(t -> Validators::validXYZ, "last name must contain characters", columns.get(4))
//);
//List<ErrorInfo> errors = specs.stream.flatMap(ValidationSpec::valdiate).collect(....);


// welche Wege stehen uns zur Verfügung
// bottom up refactoring
// top down
// mixed?

public class DataValidator {

	private CountryInfoService countryInfoService;

	public DataValidator(CountryInfoService countryInfoService) {
		this.countryInfoService = countryInfoService;
	}

	public List<ErrorInfo> check(Map<Integer, List<String>> data) {

		List<ErrorInfo> errors = new ArrayList<>();
		data.forEach((row, columns) -> {

			if (columns != null && !columns.isEmpty()) {
				{
					String value = columns.get(0);
					String message = "first name must contain characters";
					Predicate<String> validator = t -> validFirstName(t);

					List<ErrorInfo> errorInfo = validateField(row, value, message, validator);
					errors.addAll(errorInfo);
				}

				{
					String value = columns.get(1);
					String message = "last name must contain characters";
					Predicate<String> validator = t -> validLastName(t);

					List<ErrorInfo> errorInfo = validateField(row, value, message, validator);
					errors.addAll(errorInfo);
				}

				{
					String message = String.format("country code (%s) doesn't exist", columns.get(2));
					String value = columns.get(2);
					Predicate<String> validator = t -> validC(t);

					List<ErrorInfo> errorInfo = validateField(row, value, message, validator);
					errors.addAll(errorInfo);
				}

				{
					String value = columns.get(3);
					String message = String.format("birthdate (%s) can not be parsed", columns.get(3));
					Predicate<String> validator = t -> validDate(t);

					List<ErrorInfo> errorInfo = validateField(row, value, message, validator);
					errors.addAll(errorInfo);
				}

				{
					String value = columns.get(4);
					String message = String.format("income (%s) can not be parsed", columns.get(4));
					Predicate<String> validator = t -> validMoney(t);

					List<ErrorInfo> errorInfo = validateField(row, value, message, validator);
					errors.addAll(errorInfo);
				}
			}

		});
		return errors;
	}

	private boolean validFirstName(String t) {
		return validName(t);
	}

	private boolean validLastName(String t) {
		return validName(t);
	}

	private boolean validName(String value) {
		return value != null && !value.isBlank() && value.chars().allMatch(Character::isAlphabetic);
	}
	
	private List<ErrorInfo> validateField(Integer row, String value, String message, Predicate<String> validator) {
		List<ErrorInfo> errorInfo = new ArrayList<>();
		if (!validator.test(value)) {
			errorInfo.add(new ErrorInfo(row, message));
		}
		return errorInfo;
	}

	private boolean validC(String string) {
		return countryInfoService.getAllCountries().stream().anyMatch(c -> c.equals(string));
	}

	private static boolean validDate(String string) {
		try {
			LocalDate.parse(string);
		} catch (DateTimeParseException e) {
			return false;
		}
		return true;
	}

	private boolean validMoney(String s) {
		try {
			BigDecimal money = new BigDecimal(s);
			if (money.compareTo(BigDecimal.ZERO) == -1) {
				return false;
			}
			return true;
		} catch (NumberFormatException e) {
			return false;
		}
	}

}

File: csharp/DataValidator/DataValidator.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DataValidatorProject
{
    public class DataValidator
    {
        public CountryInfoService countryInfoService;

        public DataValidator(CountryInfoService countryInfoService)
        {
            this.countryInfoService = countryInfoService;
        }

        public List<ErrorInfo> Check(Dictionary<int, List<String>> data)
        {
            List<ErrorInfo> errors = new List<ErrorInfo>();
            foreach (KeyValuePair<int, List<String>> entryKey in data)
            {
                int row = entryKey.Key;
                List<string> columns = entryKey.Value;
                if (columns != null && !columns.Any())
                {
                    if (!ValidName(columns[0]))
                    {
                        errors.Add(new ErrorInfo(row, "first name must contain characters"));
                    }

                    if (!ValidName(columns[1]))
                    {
                        errors.Add(new ErrorInfo(row, "last name must contain characters"));
                    }

                    if (!ValidC(columns[2]))
                    {
                        errors.Add(new ErrorInfo(row, String.Format("country code (%s) doesn't exist", columns[2])));
                    }

                    if (!ValidDate(columns[3]))
                    {
                        errors.Add(new ErrorInfo(row, String.Format("birthdate (%s) can not be parsed", columns[3])));
                    }

                    if (InvalidMoney(columns[4]))
                    {
                        errors.Add(new ErrorInfo(row, String.Format("income (%s) can not be parsed", columns[4])));
                    }
                }

            };
            return errors;
        }

        private bool InvalidMoney(string v)
        {
            try
            {
                decimal money = Convert.ToDecimal(v);
                if (money == -1)
                {
                    return true;
                }
                return false;
            }
            catch (FormatException)
            {
                return true;
            }
        }

        private bool ValidDate(string string1)
        {
            try
            {
                Convert.ToDateTime(string1);
            }
            catch (FormatException)
            {
                return false;
            }
            return true;
        }

        private bool ValidC(string String)
        {
            List<string> allCountries = countryInfoService.getAllCountries();
            foreach (string country in allCountries)
            {
                if (country == String)
                {
                    return true;
                }
            }
            return false;
        }

        private bool ValidName(string value)
        {
            return value != null && String.IsNullOrEmpty(value) && value.Any(char.IsLetter);
        }
    }

}
